Sam readgroup might need to be fully parametrized

[Here](https://github.com/bigdatagenomics/cannoli/blob/master/core/src/main/scala/org/bdgenomics/cannoli/Bwa.scala#L113) the bwa readgroup has a hardcoded platform unit (PU) field, and the same sample name is used for the rest of the readgroup fields. 

Is the illumina PU hard coded because only illumina short reads are supported? I'm running some BGISEQ500 reads using cannoli bwa and getting some errors that I dropped in the gitter channel. I don't think there's really a difference between bgiseq500 reads and the standard illumina short reads though. 

Sample read:
```
@V100008960L3C003R006445044/1
AATTCTTACAACCTTTCGATGTGATATTTGAGCTCCCTCCAAACTCAAGTTTGTGCGTGTGTGTGTGTGTGTATCACAACCACACATTATCTACTATTTGCAGAGCTCTTGTATCAGGATATAAAAAAATCTAGGCATAAGTAATTCCAT
+
EFEFGEDGGDGEEFGGCGGGGGBDBCDGBG@GEGGFDDGGF;EFEFFEGGG@EEEGFFGBDEBEEEBFAGDDGDFB@GFC=C@DGEEBEFEEFEDDEEAC?A@5@@DCCE@C>DCFD@F?D@EGDEGFGB2D5GFB@AAE2-9ACGC;GE
```

I think ID, PU, SM, PL, and LB should be added to the BwaArgs  class. Maybe these features are remnants of flat files that you don't want though.

Ya it looks like this:

```
20/01/06 15:16:39 INFO Bwa: Piping RDDBoundFragmentDataset with 0 reference sequences, 0 read groups, and 0 processing steps to bwa with command: [bwa, mem, -t, 1, -R, @RG\tID:sample\tLB:sample\tPL:ILLUMINA\tPU:0\tSM:sample, -p, /home/james/genomic_data/reference/human_g1k_v37.fasta, -] files: []
```

Here is the driver stack trace:

```
Driver stacktrace:
20/01/06 15:17:34 INFO DAGScheduler: Job 2 failed: runJob at SparkHadoopWriter.scala:78, took 55.404197 s
20/01/06 15:17:34 ERROR SparkHadoopWriter: Aborting job job_20200106151639_0011.
org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 3.0 failed 1 times, most recent failure: Lost task 3.0 in stage 3.0 (TID 203, localhost, executor driver): org.apache.spark.SparkException: Task failed while writing rows
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:157)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:83)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.NoSuchElementException: key not found: sample
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at org.bdgenomics.adam.models.ReadGroupDictionary.apply(ReadGroupDictionary.scala:133)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:259)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:257)
	at scala.Option.foreach(Option.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:242)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.converters.AlignmentConverter.convert(AlignmentConverter.scala:242)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:741)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:738)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:131)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:129)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:141)
	... 10 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2114)
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:78)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply$mcV$sp(PairRDDFunctions.scala:1083)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopDataset(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply$mcV$sp(PairRDDFunctions.scala:1000)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopFile(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.InstrumentedPairRDDFunctions.saveAsNewAPIHadoopFile(InstrumentedPairRDDFunctions.scala:477)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$2.apply(AlignmentDataset.scala:938)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset.saveAsSam(AlignmentDataset.scala:811)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply$mcV$sp(AlignmentDataset.scala:802)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply(AlignmentDataset.scala:802)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply(AlignmentDataset.scala:802)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset.saveAsSam(AlignmentDataset.scala:801)
	at org.myome.bigbioinf.Main$.main(Main.scala:21)
	at org.myome.bigbioinf.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:845)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:161)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:184)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:86)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:920)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:929)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: org.apache.spark.SparkException: Task failed while writing rows
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:157)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:83)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.NoSuchElementException: key not found: sample
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at org.bdgenomics.adam.models.ReadGroupDictionary.apply(ReadGroupDictionary.scala:133)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:259)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:257)
	at scala.Option.foreach(Option.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:242)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.converters.AlignmentConverter.convert(AlignmentConverter.scala:242)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:741)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:738)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:131)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:129)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:141)
	... 10 more
Exception in thread "main" org.apache.spark.SparkException: Job aborted.
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:100)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply$mcV$sp(PairRDDFunctions.scala:1083)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1.apply(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopDataset(PairRDDFunctions.scala:1081)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply$mcV$sp(PairRDDFunctions.scala:1000)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsNewAPIHadoopFile$2.apply(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopFile(PairRDDFunctions.scala:991)
	at org.apache.spark.rdd.InstrumentedPairRDDFunctions.saveAsNewAPIHadoopFile(InstrumentedPairRDDFunctions.scala:477)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$2.apply(AlignmentDataset.scala:938)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset.saveAsSam(AlignmentDataset.scala:811)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply$mcV$sp(AlignmentDataset.scala:802)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply(AlignmentDataset.scala:802)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$saveAsSam$1.apply(AlignmentDataset.scala:802)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset.saveAsSam(AlignmentDataset.scala:801)
	at org.myome.bigbioinf.Main$.main(Main.scala:21)
	at org.myome.bigbioinf.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:845)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:161)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:184)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:86)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:920)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:929)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 3 in stage 3.0 failed 1 times, most recent failure: Lost task 3.0 in stage 3.0 (TID 203, localhost, executor driver): org.apache.spark.SparkException: Task failed while writing rows
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:157)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:83)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.NoSuchElementException: key not found: sample
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at org.bdgenomics.adam.models.ReadGroupDictionary.apply(ReadGroupDictionary.scala:133)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:259)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:257)
	at scala.Option.foreach(Option.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:242)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.converters.AlignmentConverter.convert(AlignmentConverter.scala:242)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:741)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:738)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:131)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:129)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:141)
	... 10 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2114)
	at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:78)
	... 39 more
Caused by: org.apache.spark.SparkException: Task failed while writing rows
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:157)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:83)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$3.apply(SparkHadoopWriter.scala:78)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.NoSuchElementException: key not found: sample
	at scala.collection.MapLike$class.default(MapLike.scala:228)
	at scala.collection.AbstractMap.default(Map.scala:59)
	at scala.collection.MapLike$class.apply(MapLike.scala:141)
	at scala.collection.AbstractMap.apply(Map.scala:59)
	at org.bdgenomics.adam.models.ReadGroupDictionary.apply(ReadGroupDictionary.scala:133)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:259)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1$$anonfun$apply$1.apply(AlignmentConverter.scala:257)
	at scala.Option.foreach(Option.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:257)
	at org.bdgenomics.adam.converters.AlignmentConverter$$anonfun$convert$1.apply(AlignmentConverter.scala:242)
	at scala.Option.fold(Option.scala:158)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.converters.AlignmentConverter.convert(AlignmentConverter.scala:242)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:741)
	at org.bdgenomics.adam.rdd.read.AlignmentDataset$$anonfun$convertToSam$2$$anonfun$10.apply(AlignmentDataset.scala:738)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:131)
	at org.apache.spark.internal.io.SparkHadoopWriter$$anonfun$4.apply(SparkHadoopWriter.scala:129)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.internal.io.SparkHadoopWriter$.org$apache$spark$internal$io$SparkHadoopWriter$$executeTask(SparkHadoopWriter.scala:141)
	... 10 more
20/01/06 15:17:34 INFO Executor: Executor killed task 8.0 in stage 3.0 (TID 208), reason: Stage cancelled
20/01/06 15:17:34 WARN TaskSetManager: Lost task 8.0 in stage 3.0 (TID 208, localhost, executor driver): TaskKilled (Stage cancelled)
20/01/06 15:17:34 INFO SparkContext: Invoking stop() from shutdown hook
20/01/06 15:17:34 INFO SparkUI: Stopped Spark web UI at http://10.1.10.183:4040
20/01/06 15:17:34 INFO MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
20/01/06 15:17:34 INFO MemoryStore: MemoryStore cleared
20/01/06 15:17:34 INFO BlockManager: BlockManager stopped
20/01/06 15:17:34 INFO BlockManagerMaster: BlockManagerMaster stopped
20/01/06 15:17:34 INFO OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
20/01/06 15:17:34 INFO SparkContext: Successfully stopped SparkContext
20/01/06 15:17:34 INFO ShutdownHookManager: Shutdown hook called
20/01/06 15:17:34 INFO ShutdownHookManager: Deleting directory /tmp/spark-8fb13355-50b6-45f4-a4bc-b2020c57a595
20/01/06 15:17:34 INFO ShutdownHookManager: Deleting directory /tmp/spark-53eda2a2-1024-4c2d-a5dc-50a73d4e855d
[M::mem_process_seqs] Processed 66668 reads in 34.200 CPU sec, 34.183 real sec
[M::process] 0 single-end sequences; 33516 paired-end sequences
[M::mem_process_seqs] Processed 66668 reads in 34.728 CPU sec, 34.705 real sec
[M::process] 0 single-end sequences; 33680 paired-end sequences
[M::mem_process_seqs] Processed 66668 reads in 35.212 CPU sec, 35.185 real sec
[M::process] 0 single-end sequences; 33430 paired-end sequences
james@james-X399-AORUS-XTREME:~/myome/bigbioinf$ [M::mem_process_seqs] Processed 66668 reads in 36.132 CPU sec, 36.112 real sec
[M::mem_process_seqs] Processed 66668 reads in 37.084 CPU sec, 37.064 real sec
[M::mem_process_seqs] Processed 66668 reads in 38.613 CPU sec, 38.602 real sec
[M::mem_process_seqs] Processed 66668 reads in 38.500 CPU sec, 38.492 real sec
^C

```
Notice at the bottom that the process throws an error and then continues to run bwa until I ctrl + c out of the shell. Everything looks normal before failure, and bwa looks like it's running as usual. System is using well below the amount of memory on my work station.

--- a/builder/src/main/java/org/bdgenomics/cannoli/builder/ContainerBuilder.java
+++ b/builder/src/main/java/org/bdgenomics/cannoli/builder/ContainerBuilder.java
@@ -17,15 +17,9 @@
  */
 package org.bdgenomics.cannoli.builder;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
 
 /**
  * Abstract container builder.

--- a/cli/src/main/scala/org/bdgenomics/cannoli/cli/Bwa.scala
+++ b/cli/src/main/scala/org/bdgenomics/cannoli/cli/Bwa.scala
@@ -20,7 +20,6 @@ package org.bdgenomics.cannoli.cli
 import grizzled.slf4j.Logging
 import htsjdk.samtools.ValidationStringency
 import org.apache.spark.SparkContext
-import org.bdgenomics.adam.models.{ ReadGroup, ReadGroupDictionary }
 import org.bdgenomics.adam.rdd.ADAMContext._
 import org.bdgenomics.adam.rdd.ADAMSaveAnyArgs
 import org.bdgenomics.adam.rdd.fragment.FragmentDataset
@@ -97,10 +96,7 @@ class Bwa(protected val args: BwaArgs) extends BDGSparkCommand[BwaArgs] with Log
       sc.loadFragments(args.inputPath, stringency = stringency)
     }
 
-    val sample = args.sample
-
     val output: AlignmentDataset = new BwaFn(args, sc).apply(input)
-      .replaceReadGroups(ReadGroupDictionary(Seq(ReadGroup(sample, sample))))
 
     val outputMaybeWithSequences = Option(args.sequenceDictionary).fold(output)(sdPath => {
       val sequences = sc.loadSequenceDictionary(sdPath)

--- a/core/src/main/scala/org/bdgenomics/cannoli/Bwa.scala
+++ b/core/src/main/scala/org/bdgenomics/cannoli/Bwa.scala
@@ -21,10 +21,11 @@ import org.apache.hadoop.fs.{ FileSystem, Path }
 import org.apache.spark.SparkContext
 import org.bdgenomics.adam.rdd.ADAMContext._
 import org.bdgenomics.adam.rdd.fragment.{ FragmentDataset, InterleavedFASTQInFormatter }
+import org.bdgenomics.adam.models.{ ReadGroup => ReadGroupModel, ReadGroupDictionary }
 import org.bdgenomics.adam.rdd.read.{ AlignmentDataset, AnySAMOutFormatter }
 import org.bdgenomics.adam.sql.{ Alignment => AlignmentProduct }
 import org.bdgenomics.cannoli.builder.CommandBuilders
-import org.bdgenomics.formats.avro.Alignment
+import org.bdgenomics.formats.avro.{ Alignment, ReadGroup }
 import org.bdgenomics.utils.cli._
 import org.kohsuke.args4j.{ Argument, Option => Args4jOption }
 import scala.collection.JavaConversions._
@@ -32,11 +33,8 @@ import scala.collection.JavaConversions._
 /**
  * Bwa function arguments.
  */
-class BwaArgs extends Args4jBase {
-  @Argument(required = true, metaVar = "SAMPLE", usage = "Sample ID.", index = 2)
-  var sample: String = null
-
-  @Args4jOption(required = true, name = "-index", usage = "Path to the BWA index to be searched, e.g. <idxbase> in bwa [options]* <idxbase>.")
+class BwaArgs extends ReadGroupArgs {
+  @Args4jOption(required = true, name = "-index", usage = "Path to the BWA index to be searched, e.g. <idxbase> in bwa [options]* <idxbase>. Required.")
   var indexPath: String = null
 
   @Args4jOption(required = false, name = "-executable", usage = "Path to the BWA executable. Defaults to bwa.")
@@ -70,7 +68,6 @@ class Bwa(
     sc: SparkContext) extends CannoliFn[FragmentDataset, AlignmentDataset](sc) {
 
   override def apply(fragments: FragmentDataset): AlignmentDataset = {
-    val sample = args.sample
 
     def getIndexPaths(fastaPath: String): Seq[String] = {
       val requiredExtensions = Seq("",
@@ -110,13 +107,15 @@ class Bwa(
       pathsWithScheme ++ optionalPathsWithScheme
     }
 
+    val readGroup = Option(args.readGroup).getOrElse(args.createReadGroup)
+
     val builder = CommandBuilders.create(args.useDocker, args.useSingularity)
       .setExecutable(args.executable)
       .add("mem")
       .add("-t")
       .add("1")
       .add("-R")
-      .add(s"@RG\\tID:${sample}\\tLB:${sample}\\tPL:ILLUMINA\\tPU:0\\tSM:${sample}")
+      .add(readGroup.toSAMReadGroupRecord().getSAMString().replace("\t", "\\t"))
       .add("-p")
       .add(if (args.addFiles) "$0" else args.indexPath)
       .add("-")
@@ -138,9 +137,11 @@ class Bwa(
     implicit val tFormatter = InterleavedFASTQInFormatter
     implicit val uFormatter = new AnySAMOutFormatter
 
-    fragments.pipe[Alignment, AlignmentProduct, AlignmentDataset, InterleavedFASTQInFormatter](
+    val alignments = fragments.pipe[Alignment, AlignmentProduct, AlignmentDataset, InterleavedFASTQInFormatter](
       cmd = builder.build(),
       files = builder.getFiles()
     )
+
+    alignments.replaceReadGroups(ReadGroupDictionary(Seq(readGroup)))
   }
 }

--- a/core/src/main/scala/org/bdgenomics/cannoli/Bwa.scala
+++ b/core/src/main/scala/org/bdgenomics/cannoli/Bwa.scala
@@ -21,13 +21,12 @@ import org.apache.hadoop.fs.{ FileSystem, Path }
 import org.apache.spark.SparkContext
 import org.bdgenomics.adam.rdd.ADAMContext._
 import org.bdgenomics.adam.rdd.fragment.{ FragmentDataset, InterleavedFASTQInFormatter }
-import org.bdgenomics.adam.models.{ ReadGroup => ReadGroupModel, ReadGroupDictionary }
+import org.bdgenomics.adam.models.ReadGroupDictionary
 import org.bdgenomics.adam.rdd.read.{ AlignmentDataset, AnySAMOutFormatter }
 import org.bdgenomics.adam.sql.{ Alignment => AlignmentProduct }
 import org.bdgenomics.cannoli.builder.CommandBuilders
-import org.bdgenomics.formats.avro.{ Alignment, ReadGroup }
-import org.bdgenomics.utils.cli._
-import org.kohsuke.args4j.{ Argument, Option => Args4jOption }
+import org.bdgenomics.formats.avro.Alignment
+import org.kohsuke.args4j.{ Option => Args4jOption }
 import scala.collection.JavaConversions._
 
 /**

--- a/core/src/main/scala/org/bdgenomics/cannoli/ReadGroupArgs.scala
+++ b/core/src/main/scala/org/bdgenomics/cannoli/ReadGroupArgs.scala
@@ -17,11 +17,10 @@
  */
 package org.bdgenomics.cannoli
 
-import org.bdgenomics.adam.models.{ ReadGroup, ReadGroupDictionary }
+import org.bdgenomics.adam.models.ReadGroup
 import org.bdgenomics.formats.avro.{ ReadGroup => ReadGroupMetadata }
 import org.bdgenomics.utils.cli._
 import org.kohsuke.args4j.{ Option => Args4jOption }
-import scala.collection.JavaConversions._
 
 /**
  * Abstract arguments that capture a read group.

--- a/core/src/main/scala/org/bdgenomics/cannoli/ReadGroupArgs.scala
+++ b/core/src/main/scala/org/bdgenomics/cannoli/ReadGroupArgs.scala
@@ -1,0 +1,87 @@
+/**
+ * Licensed to Big Data Genomics (BDG) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The BDG licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.cannoli
+
+import org.bdgenomics.adam.models.{ ReadGroup, ReadGroupDictionary }
+import org.bdgenomics.formats.avro.{ ReadGroup => ReadGroupMetadata }
+import org.bdgenomics.utils.cli._
+import org.kohsuke.args4j.{ Option => Args4jOption }
+import scala.collection.JavaConversions._
+
+/**
+ * Abstract arguments that capture a read group.
+ */
+private[cannoli] class ReadGroupArgs extends Args4jBase {
+
+  @Args4jOption(required = true, name = "-sample_id", usage = "Sample identifier for the read group. Required.")
+  var sampleId: String = null
+
+  @Args4jOption(required = false, name = "-flow_order", usage = "Read group flow order.")
+  var flowOrder: String = null
+
+  @Args4jOption(required = false, name = "-key_sequence", usage = "Read group key sequence.")
+  var keySequence: String = null
+
+  @Args4jOption(required = false, name = "-library", usage = "Read group library, defaults to -sample_id if not specified.")
+  var library: String = null
+
+  @Args4jOption(required = false, name = "-platform", usage = "Read group platform, defaults to ILLUMINA.")
+  var platform: String = null
+
+  @Args4jOption(required = false, name = "-platform_model", usage = "Read group platform model.")
+  var platformModel: String = null
+
+  @Args4jOption(required = false, name = "-platform_unit", usage = "Read group platform unit, defaults to 0.")
+  var platformUnit: String = null
+
+  @Args4jOption(required = false, name = "-predicted_median_insert_size", usage = "Read group predicted median insert size.")
+  var predictedMedianInsertSize: java.lang.Integer = null
+
+  @Args4jOption(required = false, name = "-read_group_id", usage = "Read group identifier, defaults to -sample_id if not specified.")
+  var readGroupId: String = null
+
+  @Args4jOption(required = false, name = "-run_date", usage = "Read group run date.")
+  var runDate: java.lang.Long = null
+
+  @Args4jOption(required = false, name = "-sequencing_center", usage = "Read group sequencing center.")
+  var sequencingCenter: String = null
+
+  /** Read group, to specify directly. */
+  var readGroup: ReadGroup = null
+
+  /**
+   * Create a read group from the arguments in this class.
+   *
+   * @return a read group created from the arguments in this class.
+   */
+  def createReadGroup(): ReadGroup = {
+    val builder = ReadGroupMetadata.newBuilder()
+      .setId(Option(readGroupId).getOrElse(sampleId))
+      .setLibrary(Option(library).getOrElse(sampleId))
+      .setPlatform(Option(platform).getOrElse("ILLUMINA"))
+      .setPlatformUnit(Option(platformUnit).getOrElse("0"))
+      .setSampleId(sampleId)
+
+    Option(flowOrder).foreach(builder.setFlowOrder)
+    Option(runDate).foreach(builder.setRunDateEpoch)
+    Option(predictedMedianInsertSize).foreach(builder.setPredictedMedianInsertSize)
+    Option(platformModel).foreach(builder.setPlatformModel)
+    Option(sequencingCenter).foreach(builder.setSequencingCenter)
+    ReadGroup.fromAvro(builder.build())
+  }
+}

--- a/core/src/main/scala/org/bdgenomics/cannoli/SingleEndStar.scala
+++ b/core/src/main/scala/org/bdgenomics/cannoli/SingleEndStar.scala
@@ -19,7 +19,6 @@ package org.bdgenomics.cannoli
 
 import org.apache.spark.SparkContext
 import org.bdgenomics.adam.rdd.ADAMContext._
-import org.bdgenomics.adam.rdd.fragment.FragmentDataset
 import org.bdgenomics.adam.rdd.read.{ AlignmentDataset, AnySAMOutFormatter, SAMInFormatter }
 import org.bdgenomics.adam.sql.{ Alignment => AlignmentProduct }
 import org.bdgenomics.cannoli.builder.CommandBuilders

